Aiming for release

pruzanov opened the pull request:
I think this workflow is ready to be tagged and deployed, some of the changes since the last iteration:

- added and tested baseMask parameter
- added lanes parameter
- added additionalParameters just in case
- tested with data from multiple instruments

It looks like bclconvert has no problem dealing with the data from **MiSeq**, **NovaSeq6000** and newer platforms. **Hiseq2500** and **NextSeq** tests indicate that there must be some incompatibility issues.

Tested both with dual and single barcode (indices) but only tests with **MiSeq** are included in regression tests (to keep the testing time reasonable)

lheisler committed:
added README, changelong, vidarrbuild + regression tests

Fixing before trying RT test

Fixing format

Fixing format

Improved README, fixed RT

Fixed RT

Fixed RT

Improved and corrected RT

Sorted RT setup file

Sorted RT setup file

Developing Regression Tests

pruzanov committed:
Fixing structure of samples

Shorter names

Input structure improvements

pruzanov committed:
Updating dir type, added default values for params

Previous changes were only done locally, correcting

Added RT for a single sample

Minor changes

pruzanov committed:
Support for single barcodes, more tests

Fixing identation

Fixing identation

Support for single barcodes

pruzanov committed:
Added basemask

pruzanov committed:
Added a new test, modified parameters

Minor fix

Minor fix

Minor fix

Minor fix

Minor fix

Minor fix

Minor fix

Fixing RT

Simplified RT test for debugging

Simplified RT test for debugging

Modifications to RTs

pruzanov committed:
Reverting dir type to what it was

pruzanov committed:
Reverting to larger set of RTs

Reverting to larger set of RTs

Removed NovaSeq test, takes too long (~9h)

Improved diff, ignore spaces when comparing calculated RT results

Added more tests

Removed NovaSeq test, added baseMask handling

avarsava commented on CHANGELOG.md:
```diff
@@ -0,0 +1,21 @@
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## [Unreleased] - 2024-08-02
```
Why are there two unreleased sections, and why are they dated?

pruzanov replied:
I want to separate Larry's work from mine

avarsava commented on Jenkinsfile:
```diff
@@ -0,0 +1,24 @@
+        }
+      }
+      steps {
+        sh '/.mounts/labs/gsi/vidarr/jenkins-ci-wrapper test -t /.mounts/labs/gsi/vidarr/testing-config.json'
```
@mlaszloffy is the version controlled copy of jenkins-ci-wrapper (https://bitbucket.oicr.on.ca/projects/GSI/repos/infrastructure/browse/gsici.hpc/vidarr) checked out somewhere on the cluster?

mlaszloffy replied:
`/.mounts/labs/gsi/vidarr/jenkins-ci-wrapper` is a symlink to a checkout of git/infra

avarsava commented on README.md:
```diff
@@ -1,2 +1,214 @@
+Parameter|Value|Description
+---|---|---
+`runDirectory`|String|{'description': 'Illumina run directory (e.g. /path/to/191219_M00000_0001_000000000-ABCDE).', 'vidarr_type': 'directory'}
+`runName`|String|The name of the run, this will be used for the output folder and as a file prefix
```
Is this redundant with runDirectory? (ie runName is the last level of the run directory path)

pruzanov replied:
Yes, it is the same run id

avarsava replied:
Then why does one need to provide both?

pruzanov replied:
It was implemented like this originally, for the sake of convenience, I suppose. run id is used for report name. It may certainly be simplified

avarsava commented on README.md:
```diff
@@ -1,2 +1,214 @@
+---|---|---
+`runDirectory`|String|{'description': 'Illumina run directory (e.g. /path/to/191219_M00000_0001_000000000-ABCDE).', 'vidarr_type': 'directory'}
+`runName`|String|The name of the run, this will be used for the output folder and as a file prefix
+`samples`|Array[Sample]|array of Samples, that will includes names and barcodes
```
includes -> include

What is the purpose of providing these samples? Are they not able to be discovered from within the run directory?

pruzanov replied:
We are aiming to have a workflow which can be used to process not only all samples in a run but also selected samples (for the purpose of re-running any failed processes and other edge cases). This workflow would be configured to run on a set of barcodes and produce fastq files only for a limited number of samples.

avarsava commented on README.md:
```diff
@@ -1,2 +1,214 @@
+`runDirectory`|String|{'description': 'Illumina run directory (e.g. /path/to/191219_M00000_0001_000000000-ABCDE).', 'vidarr_type': 'directory'}
+`runName`|String|The name of the run, this will be used for the output folder and as a file prefix
+`samples`|Array[Sample]|array of Samples, that will includes names and barcodes
+`modules`|String|Modules to run on hpc
```
Is the module to run not bclconvert? What are these modules used for?

pruzanov replied:
It is bclconvert, this module is loaded at runtime to perform the task

avarsava replied:
Then module should be singular and the description clarified

avarsava commented on README.md:
```diff
@@ -1,2 +1,214 @@
+#### Optional task parameters:
+Parameter|Value|Default|Description
+---|---|---|---
+`runBclconvert.firstTileOnly`|Boolean|false|Flag for processing first tile only. Default false
+`runBclconvert.noLaneSplitting`|Boolean|false|Flag to disable lane splitting. Default false
+`runBclconvert.onlyMatchedReads`|Boolean|true|Process only matched reads. Default true
+`runBclconvert.fastqCompression`|String|'gzip'|Compression type of fastq files. Default gzip
+`runBclconvert.fastqCompressionLevel`|Int|1|Fastq compression level. Default 1
+`runBclconvert.memory`|Int|32|Memory allocated for running this task. Default 32
```
"Default ___" at end of description redundant with Default column

pruzanov replied:
that is how readme generating script arranges things

avarsava replied:
readme generating script pulls from the metadata in the workflow which could drop that part of the string, or the readme can be edited after the script generates it

mlaszloffy replied:
You can remove  "Default ___" from the parameter meta (e.g. https://github.com/oicr-gsi/bclconvert/pull/1/files#diff-b3a4ca45749e1ef8f3677f5beaf1da555d059b72cf9da42432eb47d32f2b27fdR188) and leave the generation of the default column/info to the readme generator.

avarsava commented on bclconvert.wdl:
```diff
@@ -22,33 +22,39 @@ struct FastqCollection {
   }
 
   output {
-    File reports = bclconvert.reports
```
The reports have important information such as read counts, why are they being dropped?

pruzanov replied:
I was under impression that Larry removed it, but it turns out it was one of the outputs when I started working on bclconvert. I put it back

mlaszloffy replied:
We don't want the reports to be output for each bclconvert workflow run. Correct the reports have information we need, but this information should be extracted from bclconvert.reports and output as a output struct/Pair<File,Map<String,String>>/vidarr file labels

pruzanov replied:
The thing that Michael suggests is already in place. Read counts are attached to fastq output, so I am going to revert my changes and remove report from the outputs

avarsava commented:
Recommend rewriting git history

pruzanov committed:
Restoring reports as an output

Do not check md5sum for reports

Reports are not provisioned, outputs the same as from bcl2fastq

Fix for RT

pruzanov commented:
There is going to be another test added, to verify that we are able to produce and provision Unidentified reads

pruzanov committed:
Adding RT with Undetermined reads generated

Fixed metadata

avarsava commented:
- bcl2fastq adds a reads_count label per sample. How does bclconvert achieve this? It looks like it is put into outputs.json in the python block and then read in but I'd like to confirm since I'm less familiar with WDL
- could the module version be parsed out and included as a workflow label?
- could there also be a static workflow label like 'ran_on:hpc' to distinguish from on-instrument runs? (i will add this for on-instrument runs in the process that loads those runs)

pruzanov commented:
> * bcl2fastq adds a reads_count label per sample. How does bclconvert achieve this? It looks like it is put into outputs.json in the python block and then read in but I'd like to confirm since I'm less familiar with WDL
> * could the module version be parsed out and included as a workflow label?
> * could there also be a static workflow label like 'ran_on:hpc' to distinguish from on-instrument runs? (i will add this for on-instrument runs in the process that loads those runs)

1: yes, it is the same as with bcl2fastq  2: not sure if we need it (we already have workflows tagged) 3: yes, it is possible

avarsava commented:
>  2: not sure if we need it (we already have workflows tagged) 

vidarr will automatically track the 'workflow version' as vidarr sees it, which for bclconvert, will probably be '1.0.0' when we launch. this isn't the same as the actual version of bclconvert, which will look like '4.1.7' or similar. Since the on-instrument bclconvert and server bclconvert versions won't be in sync, this will be important metadata to track in the case of discrepancies

pruzanov committed:
Embedded python code moves to scripts

pruzanov committed:
Switch to two-mode workflow

Fixed RTs

Switched to two-mode workflow

avarsava commented on bclconvert.wdl:
```diff
@@ -19,189 +15,167 @@ struct FastqCollection {
-    String runName
-    Array[Sample] samples
+    String mode
+    Sample sample 
```
```suggestion
    Array[String]+ barcodes
    String name
```

pruzanov replied:
I would prefer the original implementation

avarsava replied:
Putting everything at the top level simplifies what we will need to do for the injection process. Why do we need a struct if there will only ever be one instance of it?

lheisler commented on bclconvert.wdl:
```diff
@@ -19,189 +15,167 @@ struct FastqCollection {
-  }  
 
   meta {
     author: "Lawrence Heisler"
```
expand this to include yourself Peter as the primary author

lheisler reviewed (approved):
noting that, as had been discussed

- modified to handle only a single sample from a run
- runs in both dragen and hpc mode (instead of two workflows)
- sample sheet generation moved to python in module, instead of python inline code
- undetermined reads are not generated (onlyMatchedReads = T)

this looks okay to me, but lets be sure it now conforms to what is needed for injection of data from bclconvert on-board (ie review from Mike/Alexis)

pruzanov committed:
added myself as an author

pruzanov committed:
Updated regression tests, simplified inputs

pruzanov committed:
Corrected RT for Dragen

mlaszloffy commented on vidarrtest-regression.json.in:
```diff
@@ -394,7 +394,7 @@
             "metrics_calculate": "@CHECKOUT@/tests/calculate.sh",
             "metrics_compare": "@CHECKOUT@/tests/compare.sh",
-            "output_metrics": "/.mounts/labs/gsi/testdata/bclconvert/output_metrics/MiSeqTile240108_ONESAMPLE.metrics",
+            "output_metrics": "/.mounts/labs/gsi/testdata/bclconvert/output_metrics/MiSeqTile240108_ONESAMPLE_D.metrics",
```
Is it known what is causing this difference? Just sort order or something like that...?

pruzanov replied:
I checked the files and yes, it seemed that it was the sort order. But the number of the output reads is the same in both cases, and when un-compressed the files produce identical md5sum. It appears that gzip runs differently on dragen nodes.

mlaszloffy commented on bclconvert.wdl:
```diff
@@ -211,16 +187,69 @@ task bclconvert {
+    Boolean noLaneSplitting = false
+    Boolean onlyMatchedReads = true
+    String fastqCompression = 'gzip'
+    String bclconvertScript = "/.mounts/labs/gsi/modulator/sw/Ubuntu20.04/bclconvert-scripts-1.0/bin/runBclconvert.py"
```
I see you have `runBclconvertDragen.modules` defined in the readme, but not present here. I imagine you found that modules are not support on the dragen node, as it runs Oracle Linux and our modules are built for Ubuntu 20.04.

I think this is okay for now, but can you bring this up in the workflow meeting? We should get required modules built for Oracle Linux before this workflow goes into production.

pruzanov replied:
The only module dragen uses is a couple if python scripts. These are probably Ok since it is python and not some platform-specific complied code. I agree though that a cleaner solution would be to have Oracle Linux-specific modules

mlaszloffy commented on bclconvert.wdl:
```diff
@@ -1,14 +1,10 @@
-}
-
 struct FastqFile {
     String name
```
I don't think name is required (as we're only supporting one sample) and removing it would simplify the output mapping, e.g. no need for sample name here: https://github.com/oicr-gsi/bclconvert/pull/1/files#diff-244b203ab17cb0b372f0bd5a657945a899d535bcb2bb5b3ea646ac424878a60dR66 - can it be removed please?

The output for this workflow should be `Array[Pair[File,Map[String,String]]]`

pruzanov replied:
How are we going to handle cases when we have Undefined reads provisioned? That is going to add an extra pair of files (additional sample)

mlaszloffy replied:
Maybe that should be a separate workflow. I don't think this workflow has any intention of provision out undetermined reads correct?

mlaszloffy commented on bclconvert.wdl:
```diff
@@ -19,189 +15,169 @@ struct FastqCollection {
 
   command <<<
+  set -euo pipefail
   bcl-convert -f \
```
I think you want to run this with `--bcl-only-matched-reads`:
```
  --bcl-only-matched-reads arg           For pure BCL conversion, do not output files for 'Undetermined' [unmatched] 
                                         reads (output by default)
```

Does the dragen mode need something similar too?

pruzanov replied:
Both flavours are using _--bcl-only-matched-reads_ set to true by default, not sure what you mean

pruzanov committed:
Pushing all changes adressing comments of the latest discussion

pruzanov committed:
Harmonizing RTs with new changes

pruzanov committed:
Trying struct as output

pruzanov committed:
Trying inon-empty Array of Pairs as output

pruzanov committed:
Metadata fix

pruzanov committed:
Metadata fix

pruzanov committed:
Metadata fix

avarsava commented:
Currently the output bclconvert.fastqs is recognized by vidarr as a 'files-with-labels'. If you look at the guide to vidarr output types, that means all the files get the same labels: https://github.com/oicr-gsi/vidarr/blob/master/types.md#output-types
Is this what we want? the read_count will always be the same, but we were also thinking of adding read_number as a label, which would not work in this case.

pruzanov commented:
I see that currently vidarr cannot digest the output at all, though cromwell seems to be fine with Array[Pair[File,Map[String,String]]]+  Though I think that it is Ok to apply the same label to all fastq files since it's the read number which must be the same. Not so sure about 10x data which we may want to support in a future (but it can wait until we decide that we actually want this to be supported)

avarsava commented:
>  Though I think that it is Ok to apply the same label to all fastq files since it's the read number which must be the same

There will be Read 1 and Read 2 files in the same output, won't there?

pruzanov committed:
Changing output type

pruzanov committed:
Making second reads output optional (Support for single-end reads)
